libraries/chain/block_log.cpp

```diff
@@ -477,22 +483,248 @@ namespace eosio { namespace chain {
+         }
+         return false;
+      }
+
+      bool set_active_item(uint32_t block_num) {
+         try {
+            if (active_index != npos) {
+               auto active_item = collection.nth(active_index);
+               if (active_item->first <= block_num && block_num <= active_item->second.last_block_num) {
+                  if (!log_index.is_open()) {
+                     log_index.open(active_item->second.filename_base.replace_extension("index"));
+                  }
+                  return true;
+               }
+            }
+            if (collection.empty() || block_num < collection.begin()->first )
+               return false;
+
+            auto it = --collection.upper_bound(block_num);
+
+            if (block_num <= it->second.last_block_num ) {
+               auto name = it->second.filename_base;
+               log_data.open(name.replace_extension("log"));
+               log_index.open(name.replace_extension("index"));
+               this->active_index = collection.index_of(it);
+               return true;
+            }
+            return false;
+         } catch (...) {
+            this->active_index = npos;
+            return false;
+         }
+      }
+
+      std::pair<fc::datastream<const char*>, uint32_t> datastream_for_block(uint32_t block_num) {
+         auto pos = log_index.nth_block_position(block_num - log_data.first_block_num());
+         return std::make_pair(log_data.datastream_at(pos), log_data.get_preamble().version);
+      }
+
+      void add(uint32_t start_block_num, uint32_t end_block_num, fc::path filename_base) {
+         if (this->collection.size() >= max_retained_files) {
+            auto items_to_erase = max_retained_files - this->collection.size() + 1;
+            for (auto it = this->collection.begin(); it < this->collection.begin() + items_to_erase; ++it) {
+               auto name = it->second.filename_base;
+               if (archive_dir.empty()) {
+                  // delete the old files when no backup dir is specified
+                  boost::filesystem::remove(name.replace_extension("log"));
+                  boost::filesystem::remove(name.replace_extension("index"));
+               } else {
+                  // move the the backup dir
+                  auto new_name = archive_dir / boost::filesystem::path(name).filename();
+                  boost::filesystem::rename(name.replace_extension("log"), new_name.replace_extension("log"));
+                  boost::filesystem::rename(name.replace_extension("index"), new_name.replace_extension("index"));
+               }
+            }
+            this->collection.erase(this->collection.begin(), this->collection.begin() + items_to_erase);
+            this->active_index = this->active_index == npos ? npos : this->active_index - items_to_erase;
+         }
+         this->collection.emplace(start_block_num, mapped_type{end_block_num, filename_base});
+      }
```

Comment:
Documentation indicates that inserting here could invalidate the active_index (unless it is a stable_vector, not sure how we know).  Also, do we know that the flat_map storage is fixed size per iterator, since the filename_base is not going to be guaranteed to be the same size?  I think the safest thing here is to just set active_index = npos after an add. (another corner case on top of this is that we could have an archived file that has a higher start_block_num than this start_block_num, and if we have not had any call that has changed active_index, it could be pointing at that iterator and then it would be invalid at this point)